src/array.rs

```diff
@@ -278,10 +278,10 @@ impl<T: Element, D: Dimension> PyArray<T, D> {
         }
 
         // Check if the element type matches `T`.
-        let src_dtype = arr_gil_ref.dtype();
-        let dst_dtype = T::get_dtype(ob.py());
-        if !src_dtype.is_equiv_to(dst_dtype) {
-            return Err(TypeError::new(src_dtype, dst_dtype).into());
+        let src_dtype = array.dtype();
+        let dst_dtype = T::get_dtype_bound(ob.py());
+        if !src_dtype.is_equiv_to(&dst_dtype) {
+            return Err(TypeError::new(src_dtype.into_gil_ref(), dst_dtype.into_gil_ref()).into());
```

Comment:
IIRC, `TypeError` is not part of the public API and not used elsewhere. Could you convert it as part of this PR?

Reply:
Sure, I'll do that 👍